File: app/src/main/java/be/digitalia/fosdem/viewmodels/BookmarkStatusViewModel.java
package be.digitalia.fosdem.viewmodels;

import android.app.Application;
import android.os.AsyncTask;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.arch.core.util.Function;
import androidx.core.util.ObjectsCompat;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;
import be.digitalia.fosdem.db.AppDatabase;
import be.digitalia.fosdem.model.BookmarkStatus;
import be.digitalia.fosdem.model.Event;

public class BookmarkStatusViewModel extends AndroidViewModel {

	private final AppDatabase appDatabase = AppDatabase.getInstance(getApplication());
	private final MutableLiveData<Event> event = new MutableLiveData<>();
	private final LiveData<BookmarkStatus> bookmarkStatus = Transformations.switchMap(event,
			new Function<Event, LiveData<BookmarkStatus>>() {
				@Override
				public LiveData<BookmarkStatus> apply(Event event) {
					if (event == null) {
						MutableLiveData<BookmarkStatus> singleNullResult = new MutableLiveData<>();
						singleNullResult.setValue(null);
						return singleNullResult;
					}

					return Transformations.map(
							// Prevent updating the UI when a bookmark is added back or removed back
							Transformations.distinctUntilChanged(
									appDatabase.getBookmarksDao().getBookmarkStatus(event)
							), isBookmarked -> {
								if (isBookmarked == null) {
									return null;
								}
								final boolean isUpdate = firstResultReceived;
								firstResultReceived = true;
								return new BookmarkStatus(isBookmarked, isUpdate);
							}
					);
				}
			});
	private boolean firstResultReceived = false;

	public BookmarkStatusViewModel(@NonNull Application application) {
		super(application);
	}

	public void setEvent(Event event) {
		if (!ObjectsCompat.equals(event, this.event.getValue())) {
			firstResultReceived = false;
			this.event.setValue(event);
		}
	}

	@Nullable
	public Event getEvent() {
		return event.getValue();
	}

	public LiveData<BookmarkStatus> getBookmarkStatus() {
		return bookmarkStatus;
	}

	public void toggleBookmarkStatus() {
		final Event event = this.event.getValue();
		final BookmarkStatus currentStatus = bookmarkStatus.getValue();
		// Ignore the action if the status for the current event hasn't been received yet
		if (event != null && currentStatus != null && firstResultReceived) {
			AsyncTask.SERIAL_EXECUTOR.execute(() -> {
				if (currentStatus.isBookmarked()) {
					appDatabase.getBookmarksDao().removeBookmark(event);
				} else {
					appDatabase.getBookmarksDao().addBookmark(event);
				}
			});
		}
	}
}

File: app/src/main/java/be/digitalia/fosdem/viewmodels/BookmarkStatusViewModel.kt
package be.digitalia.fosdem.viewmodels

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import be.digitalia.fosdem.alarms.AppAlarmManager
import be.digitalia.fosdem.db.BookmarksDao
import be.digitalia.fosdem.flow.stateFlow
import be.digitalia.fosdem.flow.versionedResourceFlow
import be.digitalia.fosdem.model.BookmarkStatus
import be.digitalia.fosdem.model.Event
import be.digitalia.fosdem.utils.BackgroundWorkScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.flatMapLatest
import kotlinx.coroutines.flow.flowOf
import kotlinx.coroutines.launch
import javax.inject.Inject

@HiltViewModel
class BookmarkStatusViewModel @Inject constructor(
    private val bookmarksDao: BookmarksDao,
    private val alarmManager: AppAlarmManager
) : ViewModel() {

    private val eventStateFlow = MutableStateFlow<Event?>(null)

    @OptIn(ExperimentalCoroutinesApi::class)
    val bookmarkStatus: StateFlow<BookmarkStatus?> =
        stateFlow(viewModelScope, null) {
            eventStateFlow.flatMapLatest { event ->
                if (event == null) {
                    flowOf(null)
                } else {
                    versionedResourceFlow(bookmarksDao.version) {
                        val isBookmarked = bookmarksDao.getBookmarkStatus(event)
                        BookmarkStatus(event.id, isBookmarked)
                    }
                }
            }
        }

    var event: Event?
        get() = eventStateFlow.value
        set(value) {
            eventStateFlow.value = value
        }

    fun toggleBookmarkStatus() {
        val event = eventStateFlow.value
        val currentStatus = bookmarkStatus.value
        // Ignore the action if the status for the current event hasn't been received yet
        if (event != null && currentStatus != null && event.id == currentStatus.eventId) {
            if (currentStatus.isBookmarked) {
                removeBookmark(event)
            } else {
                addBookmark(event)
            }
        }
    }

    private fun removeBookmark(event: Event) {
        val eventIds = longArrayOf(event.id)
        BackgroundWorkScope.launch {
            if (bookmarksDao.removeBookmarks(eventIds) > 0) {
                alarmManager.onBookmarksRemoved(eventIds)
            }
        }
    }

    private fun addBookmark(event: Event) {
        BackgroundWorkScope.launch {
            bookmarksDao.addBookmark(event)?.let { alarmInfo ->
                alarmManager.onBookmarksAdded(listOf(alarmInfo))
            }
        }
    }
}